release-21.2: kv: alter error for SET TRANSACTION AS OF SYSTEM TIME

e-mbrown opened the pull request:
Backport 1/1 commits from #79090.

/cc @cockroachdb/release

---

if reads or writes  are already performed

Resolves #77265 

When a txn performed a read or write before setting up a historical timestamp a crash report was generated. This commit handles the error case.

Release note: None
Release justification: fix an internal error

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82373)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82373)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @e-mbrown and @rafiss)

___
*[`pkg/kv/sender.go` line 335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82373#-N3bGyDABDcc7ADWE9zx:-N3bGyDBAEucnZmHOm-u:b-be0ny0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4892e00a8bc1dfca692fafa24439b68ffdc03f21/pkg/kv/sender.go#L335)):*
> ```Go
> 
> 	// ClearTxnRetryableErr clears the retryable error, if any.
> 	ClearTxnRetryableErr(ctx context.Context)
> ```

your backport should not include these two functions (GetTxnRetryableErr and ClearTxnRetryableErr) -- since that would be adding more functionality than what your original PR was doing


<!-- Sent from Reviewable.io -->

e-mbrown committed:
kv: alter error for SET TRANSACTION AS OF SYSTEM TIME
if reads or writes  are already performed

When a txn performed a read or write before setting up a historical timestamp a crash
report was generated. This commit handles the error case.

Release note: None
